### PR TITLE
Towards ADTs

### PR DESCRIPTION
**Pull request by buggymcbugfix**
# Status

Comments welcome

- [x] Add pattern language to AST
- [x] Rename `enum` to `type`
- [x] Parsing
- [x] Constructor parameter names
- [x] Duplicate checks
  - [x] field names
  - [x] constructor names
  - [x] type definition names
- [x] Type check constructor parameters
- [ ] Basic (neither sound nor complete) exhaustivity check (warning)
- [x] Implement evaluation

# What we are adding to FCL
## Named tuples

As a first step towards ADTs, we allow sums of named tuples. Type parameters are left for a future PR.

```c
type Size { Big; Small }

type Item {
  Chips(Size size);
  Milkshake(int millilitres, bool sprinkles);
  Salad;
}
```

Giving names to the type parameters in the `type` declaration serves for documentation and to give better errors, e.g.: for `Milkshake("blue", True)` we can report

> Type error: Expected `int` in `millilitres` field of `Milkshake`, but got `text`

In future we may want to allow non-positional arguments (also in method calls), e.g.: `Milkshake(sprinkles = True, millilitres = 1)`.

## Generalising `case`

At the moment we are allowing `case` only on enums. Since we now have a proper pattern language, we can extend this to allow arbitrary expressions, e.g.:

``` c
foo(int m) {
  case (m) {
    0 -> "none";
    1 -> "one";
    _ -> "tons"
  }
}
```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,4 +1,5 @@
 <p align="center">
+=======
```
Comment (buggymcbugfix) [inline README.md]: whoops

**Comment by buggymcbugfix** [comment]
@sdiehl, @Acentelles this is ready for review. I won't be back until 20 June, so feel free to make any minor changes if you want to merge this asap.

**Comment by buggymcbugfix** [comment]
PS: the SDK will need changing because we now have `VConstr : NameUpper -> [Value] -> Value`.

**Comment by sdiehl** [comment]
:tada: Great work @buggymcbugfix .

File: tests/Reference.hs [hunk 2]
```diff
@@ -9,7 +9,6 @@ Test fixtures.
 {-# LANGUAGE OverloadedStrings #-}
 {-# LANGUAGE RecordWildCards #-}
 {-# LANGUAGE TypeApplications #-}
-{-# LANGUAGE TemplateHaskell #-}
```
Comment (sdiehl) [inline tests/Reference.hs]: :clap:

File: src/Language/FCL/Storage.hs [hunk 3]
```diff
@@ -168,7 +168,7 @@ instance FromJSON Value where
         "VTimeDelta" -> VTimeDelta  <$> o .: "contents"
         "VSig"      -> VSig      <$> o .: "contents"
         "VText"     -> VText     <$> o .: "contents"
-        "VEnum"     -> VEnum     <$> o .: "contents"
+        "VADT"     -> VConstr   <$> o .: "name" <*> o .: "args"
```
Comment (Acentelles) [inline src/Language/FCL/Storage.hs]: @buggymcbugfix the tags of VConstr used in its json serialization/deserialization don't match. Which one do you want to use?

Comment (buggymcbugfix) [reply src/Language/FCL/Storage.hs]: Good spot! Use this fix: https://github.com/adjoint-io/fcl/pull/13/commits/547618d57f373bc013d494af261a9dea343f936a

File: src/Language/FCL/Typecheck.hs [hunk 4]
```diff
@@ -274,74 +281,74 @@ bracketScopeTmpEnvM thingToDo = do
...
-    validateTypes expected actual =
-      when (not $ validMethodArgType expected actual) $
-        Left $ InvalidArgType (locVal $ methodName method) expected actual
-
-    validMethodArgType :: Type -> Type -> Bool
-    validMethodArgType (TAsset _) (TAsset TAny) = True
-    validMethodArgType (TAsset TAny) (TAsset _) = True
-    validMethodArgType (TNum np1) (TNum np2) = np1 <= np2
-    validMethodArgType t1 t2 = t1 == t2
+-- tcMethodCall :: ADTInfo -> Method -> [Value] -> Either TypeErrInfo ()
```
Comment (Acentelles) [inline src/Language/FCL/Typecheck.hs]: @buggymcbugfix why is this function commented out?

**Pull request merged**